static/app/components/forms/compactSelect.tsx

```diff
@@ -178,41 +178,51 @@ function CompactSelect<OptionType extends GeneralSelectValue = GeneralSelectValu
   // Keep an internal copy of the current select value and update the control
   // button's label when the value changes
   const [internalValue, setInternalValue] = useState(valueProp ?? defaultValue);
-  const [label, setLabel] = useState(getLabel(valueProp ?? null));
 
   // Update the button label when the value changes
-  function getLabel(newValue): React.ReactNode {
+  const getLabel = useCallback((): React.ReactNode => {
+    const newValue = valueProp ?? internalValue;
     const valueSet = Array.isArray(newValue) ? newValue : [newValue];
-    const optionSet = valueSet
+    const selectedOptions = valueSet
       .map(val => getSelectedOptions<OptionType>(options, val))
       .flat();
-    const firstOptionLabel = optionSet[0]?.label ?? '';
 
     return (
       <Fragment>
-        <ButtonLabel>{firstOptionLabel}</ButtonLabel>
-        {optionSet.length > 1 && <StyledBadge text={`+${optionSet.length - 1}`} />}
+        <ButtonLabel>{selectedOptions[0]?.label ?? ''}</ButtonLabel>
+        {selectedOptions.length > 1 && (
+          <StyledBadge text={`+${selectedOptions.length - 1}`} />
+        )}
       </Fragment>
     );
-  }
+  }, [options, valueProp, internalValue]);
 
+  const [label, setLabel] = useState<React.ReactNode>(null);
   useEffect(() => {
-    const newValue = valueProp ?? internalValue;
-    const newLabel = getLabel(newValue);
-    setLabel(newLabel);
-  }, [valueProp ?? internalValue, options]);
+    setLabel(getLabel());
+  }, [getLabel]);
 
-  // Calculate & update the trigger button's width, to be used as the
-  // overlay's min-width
+  // Calculate the current trigger element's width. This will be used as
+  // the min width for the menu.
   const [triggerWidth, setTriggerWidth] = useState<number>();
+  // Update triggerWidth when its size changes using useResizeObserver
+  const updateTriggerWidth = useCallback(async () => {
+    // Wait until the trigger element finishes rendering, otherwise
+    // ResizeObserver might throw an infinite loop error.
+    await new Promise(resolve => window.setTimeout(resolve));
+
+    const newTriggerWidth = triggerRef.current?.offsetWidth;
+    newTriggerWidth && setTriggerWidth(newTriggerWidth);
+  }, [triggerRef]);
+  useResizeObserver({ref: triggerRef, onResize: updateTriggerWidth});
+  // If ResizeObserver is not available, manually update the width
+  // when any of [trigger, triggerLabel, triggerProps] changes.
```

Comment:
same code as in https://github.com/getsentry/sentry/blob/vl/chart-option-selector/static/app/components/dropdownMenuControlV2.tsx#L125-L145